fs/grub/disk_io.c

```diff
@@ -128,30 +128,31 @@ int filemax;
 int
 rawread (int drive, int sector, int byte_offset, int byte_len, char *buf)
 {
+	/* FIXME: Simplify this function, as BootIdeReadSector() supports offset and buffering */
 	sector+=byte_offset/512;
 	byte_offset%=512;
 	
 	while(byte_len) {
-		int nThisTime=512;
-	       	if(byte_len<512) nThisTime=byte_len;
+		int nThisTime = (byte_len < 512 ? byte_len : 512);
+
 		if(byte_offset) {
 			u8 ba[512];
-			if(BootIdeReadSector(drive, buf, sector, 0, 512)) {
-				bprintf("Unable to get first sector\n");
-                             	return 1;
+			if(BootIdeReadSector(drive, &ba, sector, 0, 512)) {
+				bprintf("Unable to read sector %d of IDE drive %d (512 bytes)\n", sector, drive);
+				return 1;
+			}
+			/* FIXME: Uncommenting next line breaks ReactOS boot from native FAT32 partition */
+			//nThisTime -= byte_offset;
```

Comment:
Why would this cause issues?
Not having it can cause out-of-bounds reads.

Sounds like an error elsewhere.

Reply:
Yes probably there is some other error elsewhere. See the notes from PR description.

Reply:
This is really hard to accept. The Discord discussion is not linked, so this is without context and this touches code, has obvious issues, but is still supposed to be merged.

I'm willing to accept this under the following conditions:

- Only if this is read-only.
- This is using a moderately good IDE driver which stays read-only (won't corrupt peoples drives).
- If an issue with more context is created after merge.

Reply:
- Yes, it's read-only.
- It seems like yes, driver is used in read-only mode.
- Issue will be created.

Reply:
I'm against merging this particular piece of code as it is. The logic is just
flawed and having the commented-out line in the code causes more confusion.

Consider this case:

```c
byte_offset = 510
nThisTime = 10

BootIdeReadSector(drive, ba, sector, 0, 512);
nThisTime -= byte_offset; // <-- a
memcpy(out_buf, &ba[byte_offset], nThisTime)
```

If the line (a) is left in, now we get a negative value in `nThisTime`! If we
take the line out, then we get our 10 bytes, but we overread the block buffer
and start reading junk from the program stack. I guess we have gotten lucky in
testing with it.

Without getting into a total refactor of this code base, please consider this
alternative logic for now, which I think should meet the demands:

```c
// Possibly defined elsewhere
#define SECTOR_SIZE 512
#define MIN(x,y) ((x)<(y)?(x):(y))

int rawread(int drive, int sector, int byte_offset, int byte_len, char *buf)
{
	u8 sector_buf[SECTOR_SIZE];
	int status;
	int bytes_to_copy;

	// Advance sector counter by offset to get sub-sector-size offset
	sector += byte_offset/SECTOR_SIZE;
	byte_offset %= SECTOR_SIZE;

	// Handle first sector which may have an offset
	if (byte_offset) {
		// Determine max number of bytes to copy from this sector
		bytes_to_copy = MIN(byte_len, SECTOR_SIZE-byte_offset);

		// Read entire sector into temporary buffer
		status = BootIdeReadSector(drive, sector_buf, sector, 0, SECTOR_SIZE);
		if (status) goto error;

		// Copy sector data into dest buf
		memcpy(buf, &sector_buf[byte_offset], bytes_to_copy);
		buf += bytes_to_copy;
		byte_len -= bytes_to_copy;
		sector++;
	}

	// We are now sector-aligned. Read remainder directly into destination buffer
	while (byte_len > 0) {
		bytes_to_copy = MIN(byte_len, SECTOR_SIZE);
		status = BootIdeReadSector(drive, buf, sector, 0, bytes_to_copy);
		if (status) goto error;
		
		buf += bytes_to_copy;
		byte_len -= bytes_to_copy;
		sector++;
	}

	// I'm unsure about this return value, but have retained it from original
	// code
	return !(errnum);

error:
	bprintf("Unable to read sector %d of IDE drive %d (%d bytes)\n", sector, drive, bytes_to_copy);
	return 1;
}
```

Reply:
@mborgerson your function worked perfectly, many thanks! I replaced it.